README.md
# Bootstrap-creator

Hay ocasiones que estamos creando constantemente proyectos con bootstrap y no queremos tener bower o parecidos para instalar un simple bootstrap. Con este script podremos crear una página .html con boostrap integrado (archivos css y js incluidos). Además podemos escoger por parámetros si queremos que nuestro .html apunte a los archivos bootstrap CDN o a los archivos locales.
También podemos crear simplemente un .html con una plantilla básica de bootstrap (en el caso que ya tengamos los archivos bootstrap)

Ejemplo para ejecutar el script
```shell
strap
```

## Instalación

La instalación es muy básica, añade un alias a tu .bashrc

```shell
git clone https://github.com/victormln/bootstrap-creator.git
cd bootstrap-creator
./install.sh
```

## Configuración

Se pueden configurar varios parámetros. Por ejemplo que no busque actualizaciones automáticas, modificar el editor con el que se abrirá tu proyecto creado y otras configuraciones. Si se abre el archivo **user.conf** se podrán acceder a todas las configuraciones.

## Uso

Este script se puede ejecutar de varias maneras, una de ellas (y la más simple):
```shell
strap proyecto/test
```

Esto te creará en el directorio actual, una carpeta llamada proyecto y dentro otra llamada test.

Todos los argumentos disponibles:

|Argumento           |Abreviado|Significado                                   |Uso|
| ------------- | ---- | ---------------------------------------- |----------|
|`--help`     | -h | Abre/edita el archivo de configuración del script  |`strap --help`      |
|`proyecto`       |     | Se creará la carpeta proyecto y dentro un .html y los archivos bootstrap necesarios        |`strap proyecto/testing`  |
|`-file`     |  | Creará tan solo un archivo con los links a los css y js de Bootstrap (no creará ni una carpeta, ni te pondrá los archivos css/js).  |`strap -f test.html`      |
|`--conf`     |  | Abre/edita el archivo de configuración del script  |`strap --conf`      |
|`--cdn`     |  | Busca actualizaciones disponibles.  |`gp proyecto --cdn`      |
|`--title`     |  | Usará los links a los css y js de CDN.  |`strap proyecto --title "Hola mundo"`      |
|     |`-v`  | Muestra la versión instalada del script.  |`gp -v`      |


changelog.txt
#############
# Changelog #
#############

* Version: 1.0.0 *
#----------------#

- Añadido bootstrap min 3.3.7


update.sh
#!/bin/bash
# Fitxer: update.sh
# Autor: Víctor Molina Ferreira (victor)
# Data: 17/01/2017
# Versión: 1.0

#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.

#  Descripción: Comprueba si el script está a la ultima version

# Compruebo que sistema está usando para hacer ping
# Si es Linux o Mac
if [ "$(uname -s)" == "Linux" ] || [ "$(uname)" == "Darwin" ]; then
    ping -c 1 www.google.com &> /dev/null
    has_internet=$(echo $?)
    # Si es Windows
elif [ "$(expr substr $(uname -s) 1 10)" == "MINGW32_NT" ] ||
  [ "$(expr substr $(uname -s) 1 10)" == "MINGW64_NT" ]; then
    ping -n 1 www.google.com > /dev/null
    has_internet=$(echo $?)
fi

# Si el ping se ha realizado correctamente es que tiene internet
# por lo que se buscaran actualizaciones
if [ $has_internet -eq 0 ]
then
  if $search_ota
  then
    tieneUltimaVersion=false
    # Conseguimos la ultima version que hay en github y le quitamos los puntos
    ultimaVersion=$(curl -s https://raw.githubusercontent.com/victormln/bootstrap-creator/master/user.conf | tail -1 | cut -d'=' -f 2) > /dev/null
    ultimaVersionSinPuntos=$( echo $ultimaVersion | tr -d ".")
    # Miramos que versión tiene el usuario actualmente
    versionActualSinPuntos=$(cat $( dirname "${BASH_SOURCE[0]}" )/user.conf | tail -1 | cut -d'=' -f 2 | tr -d ".")
    # Comprobamos si la versionActual es igual o mas grande que la ultimaVersion
    # es igual a la versionActual.
    if (( $versionActualSinPuntos>=$ultimaVersionSinPuntos ))
    then
    	tieneUltimaVersion=true
    else
    	directorioActual=$(pwd)
    	# Nos colocamos en el directorio del script, para actualizarlo
    	cd "$( dirname "${BASH_SOURCE[0]}" )"
    	cd ..
    	# Mostramos el mensaje de que hay una nueva actualización
    	echo "###########################################"
    	echo -e "${WARNING}¡NUEVA ACTUALIZACIÓN!${NC}"
    	echo "Tienes la versión: $version"
    	echo "Versión disponible: $ultimaVersion"
    	echo "###########################################"
    	# Si tiene las actualizaciones automaticas, no se le pide nada
    	if $automatic_update
    	then
    		# Si es así, hacemos un pull y le actualizamos el script
    		echo "Hay una nueva actualización y tienes activadas las descargas automáticas."
    		git pull | tee >(echo "Actualizando... Por favor, espere ...")
    		echo -e "${OK}[OK] ${NC}La actualización ha acabado, por favor, vuelva a iniciar el script.";
    	else
    	  echo "Hay una nueva versión de este script y se recomienda actualizar."
    	  echo "Quieres descargarla y así tener las últimas mejoras? y/n o s/n"
    	  # Preguntamos si quiere actualizar
    	  read actualizar
    	  if [ $actualizar == "s" ] || [ $actualizar == "y" ]
    	  then
    			directorioActual=$(pwd)
    			cd "$( dirname "${BASH_SOURCE[0]}" )"
    			cd ..
    	    # Si es así, hacemos un pull y le actualizamos el script
    	  	git pull | tee >(echo "Actualizando... Por favor, espere ...")
          # Iniciamos de nuevo el script para ejecutar el script actualizado
          exec $( dirname "${BASH_SOURCE[0]}" )/bootstrap-creator.sh
    			cd $directorioActual
    			echo -e "${OK}[OK] ${NC}La actualización ha acabado, por favor, vuelva a iniciar el script.";
          exit
    	  else
    	    # En el caso que seleccione que no, muestro un mensaje.
    	    echo -e "${WARNING}¡AVISO!${NC} NO se actualizará (aunque se recomienda)."
    			# Damos por su puesto que tiene la ultima version,
    			# para que el script no entre en bucle
    			tieneUltimaVersion=true
    	  fi
    	fi
    	# Cambiamos al directorio donde el usuario tiene sus cambios
    	cd $directorioActual
    fi
  fi
else
  echo -e "${WARNING}[AVISO] ${NC}No tienes internet. Para buscar actualizaciones se necesita internet."
fi


bootstrap-creator.sh
#!/bin/bash
# Fitxer: bootstrap-creator.sh
# Autor: Víctor Molina Ferreira (victor)
# Data: 17/01/2017
# Versión: 1.0

#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.

#  Descripción: Subida de los cambios a un repositorio git (ejecutar el script donde se tenga el .git)

# Mensajes de color
ERROR='\033[0;31m'
OK='\033[0;32m'
WARNING='\033[1;33m'
NC='\033[0m'
BLUE='\033[0;34m'
PURPLE='\033[0;35m'
ORANGE='\033[0;33m'
CYAN='\033[0;36m'

# Cogemos los datos del archivo .conf
source $( dirname "${BASH_SOURCE[0]}" )/user.conf

# Si están activadas las actualizaciones automáticas
# Doy permiso al update.sh
chmod +x $( dirname "${BASH_SOURCE[0]}" )/update.sh
# Comprobaré si hay alguna versión nueva del programa autopush
# y lo mostraré en pantalla
#source $( dirname "${BASH_SOURCE[0]}" )/update.sh

function showHelp {
  echo -e "usage: bootstrap-creator name_of_project [--title] [--cdn] - Script que te permite poner un bootstrap los archivos necesarios para tener un bootstrap en un solo comando."

  echo -e "\n${CYAN}name_of_project${NC}"
  echo -e "\tEs donde se guardará el bootstrap. Este parámetro es de uso obligatorio."
  echo -e "\tTambién puedes añadir una ruta."

  echo -e "\n${CYAN}[--title] nombre_del_title${NC}"
  echo -e "\tSi pones este parámetro seguido de un nombre de título, te pondrá ese título entre los tags <title> del html"
  echo -e "\tSi no pones este parámetro, el tag <title> estará vacío."

  echo -e "\n${CYAN}[--cdn]${NC}"
  echo -e "\tSi pones este parámetro, en vez de los archivos locales de css y js, tendrás un enlace a los archivos en CDN"
  echo -e "\tSi no pones este parámetro, se te pondrán los archivos de bootstrap localmente."

}

# Hacer con array associativo
function parseOption {
  if [ -z $1 ]
  then
    echo -e "${ERROR}[ERROR] ${NC}Por favor, mínimo tiene que poner un parámetro."
    echo "Puedes ver la ayuda del scrip con --help"
  else
    if [ "$1" == "--help" ]
    then
      showHelp
    else
      parseTitle $2 $3
      mkdir -p $1
      touch $1/index.html
      cp $( dirname "${BASH_SOURCE[0]}" )/bootstrap-files/template.html $1/index.html
      fillTemplate $1
      copyDate $1
      echo -e "${OK}[OK] ${NC}Se ha creado tu proyecto $1 correctamente."
      if $open_editor
      then
        if command -v $text_editor $1 >/dev/null 2>&1
        then
          $text_editor $1 &
        else
          echo -e "${ERROR}[ERROR] ${NC}Por favor, ponga un editor que suela usar en el archivo user.conf."
          echo "Por ejemplo: text_editor=atom"
          echo "Se va a abrir por defecto el index.html creado con gedit"
          gedit $1/index.html &
        fi
      fi
    fi
  fi
}

function fillTemplate {
  if $cdn
  then
    sed "/<\/title>/ r $( dirname "${BASH_SOURCE[0]}" )/bootstrap-files/header_cdn.html" $1/index.html > $1/tmp.html
    mv $1/tmp.html $1/index.html
  else
    sed "/<\/title>/ r $( dirname "${BASH_SOURCE[0]}" )/bootstrap-files/header_local.html" $1/index.html > $1/tmp.html
    mv $1/tmp.html $1/index.html
  fi
  #
  if [ -z $title ]
  then
    sed -i 's/<titulo>//g'  $1/index.html
  else
    sed -i "s/<titulo>/$title/g" $1/index.html
  fi
}

function copyDate {
  cp -R $( dirname "${BASH_SOURCE[0]}" )/bootstrap-files/$version_bootstrap/* $1
}

# Dados dos argumentos --title y "Nombre del <title>"
function parseTitle {
  if [ "$1" == "--title" ]
  then
    project_name=$(echo "$2" | grep "^--")
    if [ -z $project_name ]
    then
      title=$2
    else
      echo -e "${ERROR}[ERROR] ${NC}Por favor, si pone el parámetro --title, le hará falta poner el título de la página."
      echo "Si no quiere ningún title por defecto, no ponga el parámetro --title"
      echo "Uso: --title \"Titulo del <title>\""
      exit
    fi
  fi
}

function cdn {
  for var in "$@"
  do
    if [ "$var" == "--cdn" ]
    then
      cdn=true
    fi
  done
}
IFS=''
directorio_actual=$(pwd)
title=""
cdn $*
parseOption $*
